rfd/0173-terraform-machine-id.md

```diff
@@ -0,0 +1,303 @@
+
+#### The existing advanced user
+
+> I am an existing Teleport user who uses Terraform to manage my Teleport resources.
+> I want to benefit from MFA4A while keeping my Terraform pipeline working.
+> My Terraform code is commited in a git repo and applied in CI pipelines, either via a CI runner (GitHub
+> Actions/GitLab CI) or via a dedicated service (Terraform Cloud/Spacelift).
+
+The user is expected to know the Teleport basics and be able to perform advanced setup actions such as
+creating roles and bot resources for MachineID. Some setup complexity is not an issue for this persona.
+
+#### The "getting started" users
+
+We have two personas:
+
+- The prospect:
+> I am not yet a Teleport user and want to test what the Teleport IaC experience looks like before using/buying Teleport.
+> IaC will be mandatory for me because I am subject to various internal policies and I must check if Teleport is
+> compatible with my IaC policies.
+> I have prior Terraform experience.
+
+- The existing manual user
+> I am an existing Teleport user that have been managing Teleport resources manually.
+> Today I am investigating how to make my setup more robust by leveraging an IaC tool.
+> I might not have prior Terraform experience.
+
+Both personas don't have (yet) dedicated infrastructure such as CI runners to deploy the Terraform code.
+They want to prototype and validate everything works using the local computer.
+
+Both personas have very little knowledge about Teleport and don't know what MachineID is and how to configure it.
+We want those users to succeed as fast as possible without having to read all the MachineID docs. When they will go
+to production, they will hopefully turn into [existing advanced users](#the-existing-advanced-user) and setup CI/CD
+pipelines for their deployment.
+
+### Implementation
+
+We will introduce two new mechanisms in the Terraform provider:
+- MachineID joining
+- Resource bootstrapping
+
+#### MachineID joining
+
+The Terraform provider will be able to take a MachineID token and join method to join the Teleport cluster without the
+need to configure and run a sidecar tbot process.
+
+The Terraform provider will run an in-process `tbot` instance and generate a client with automatic cert-renewal.
+The `tbot` instance will not persist its certificates on the disk.
+
+A typical GitHub actions provider configuration would look like:
+
+```hcl
+provider "teleport" {
+  addr               = "mytenant.teleport.sh:443"
+  onboarding = {
+    token       = "gha-runner"
+    join_method = "github"
+  }
+}
+```
+
+As Teleport does not support yet reusable bot tokens so the `token` join method will not be supported.
+
+Implementation-wise, this would reuse the
+same [in-process tbot library](https://github.com/gravitational/teleport/tree/rfd/173-terraform-machine-id/integrations/operator/embeddedtbot)
+we used for the Teleport operator.
+
+#### Resource bootstrapping
```

Comment:
The issue is that the user has to create a new bot for every Terraform invocation because bots cannot be reused. The user would have to generate a secret token, put in a TF file, then run TF, which will delete the token. Re-running the same command would then fail. Also if the command is not run, we end up with secret tokens in the TF code, which can then be committed (we can workaround this bit by reading the token from a file).